app/uk/gov/hmrc/incometaxsessiondata/services/SessionService.scala

```diff
@@ -16,30 +16,46 @@
 
 package uk.gov.hmrc.incometaxsessiondata.services
 
-import uk.gov.hmrc.incometaxsessiondata.models.{Session, SessionData}
+import uk.gov.hmrc.incometaxsessiondata.models.{FullDuplicate, PartialDuplicate, NonDuplicate, Session, SessionData, SessionDataRequest, SessionDuplicationType}
 import uk.gov.hmrc.incometaxsessiondata.repositories.SessionDataRepository
 
 import javax.inject.{Inject, Singleton}
 import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
 
 @Singleton
 class SessionService @Inject() (
-  repository: SessionDataRepository
+  val repository: SessionDataRepository
 )(implicit ec: ExecutionContext) {
 
-  def get(sessionId: String): Future[Either[Throwable, Option[SessionData]]] =
-    repository.get(sessionId) map {
+  def get(request: SessionDataRequest[_]): Future[Option[SessionData]] = {
+    repository.get(request) map {
       case Some(data: Session) =>
-        Right(Some(data))
-      case None                => Right(None)
+        Some(SessionData.fromSession(data))
+      case None => None
     }
+  }
 
-  def set(sessionData: SessionData): Future[Boolean] =
-    repository.set(sessionData)
+  def set(session: Session): Future[Either[Throwable, Boolean]] = {
+    Try {
```

Comment:
we would expect records search and duplication handling to be on the level of service